Add autonavi.com (China) map layers

magicbug2013 opened the pull request:
为IITC增加高德地图，目前修正了portal、link、field及 player tracker plugin的坐标，显示正常，但原来的收藏的portal位置将会有偏移，需要手工重新添加一次。另由于苹果机上的iitc-mobile在新版IITC不能选择图层了，插件在启用后将会将把用户的默认地图切换成高德路网。

magicbug2013 committed:
Create basemap-gaode.user.js

magicbug2013 committed:
Update basemap-gaode.user.js

magicbug2013 committed:
Update basemap-gaode.user.js

magicbug2013 committed:
Update basemap-gaode.user.js

去掉设置默认图层功能，如果使用 iitc-mobile请手动开启。

johnd0e commented:
>  Because iitc-mobile on iPhone can't select the layer in the new version of IITC

@magicbug2013 Shoudn't you open issue? May be it could be fixed.

> please remove the following comments before adding to user script

Is it possible to detect ios programmatically?

> 但原来的收藏的portal位置将会有偏移，需要手工重新添加一次。

Are you talking about Bookmarks?  (Google translated as 'original collection's portal location')

magicbug2013 commented:
@johnd0e 
> Is it possible to detect ios programmatically?

It seems impossible to debug.
>Are you talking about Bookmarks? (Google translated as 'original collection's portal location')

yes, it means Bookmarks, because Gaode, Google, Tencent and Baidu maps are offset in mainland China.

johnd0e commented:
> It seems impossible to debug.

Well, I'm sure that it's possible to use something like [this](https://stackoverflow.com/questions/9038625/detect-if-device-is-ios).
(Or even something more advanced, like [mobile-detect.js](http://hgoebl.github.io/mobile-detect.js/))

But better would be just fix iPhone issue (if possible), because it concerns all layers, not just this.

modos189 commented:
About the adjustment of coordinates. Maybe add this to IITC? Then you need a way for the plugin to report the need to enable the adjustment.

johnd0e commented:
> Then you need a way for the plugin to report the need to enable the adjustment.

IITC could provide general transform function, and any plugin be able to use it.

johnd0e commented:
And I'm still thinking about Bookmarks etc..
May it be possible to leave coordinates intact, and just offset tiles screen position?
(just brainstorming)

magicbug2013 commented:
> About the adjustment of coordinates. Maybe add this to IITC? Then you need a way for the plugin to report the need to enable the adjustment.

That's great. That the plugins only needs to define an offset to move the map layer.

magicbug2013 commented:
> And I'm still thinking about Bookmarks etc..
> May it be possible to leave coordinates intact, and just offset tiles screen position?
> (just brainstorming)

yes, actually Bookmarks only needs to use portal's guid, it does not need to use latlngs.

magicbug2013 committed:
....a small problem

magicbug2013 commented:
> > And I'm still thinking about Bookmarks etc..
> > May it be possible to leave coordinates intact, and just offset tiles screen position?
> > (just brainstorming)
> 
> yes, actually Bookmarks only needs to use portal's guid, it does not need to use latlngs.

I'm wrong. We have to consider how to display bookmarks when protal is not displayed.

modos189 commented:
> IITC could provide general transform function, and any plugin be able to use it.

Yes, IITC could provide general transform function, but it should affect other plugins. So, we need to make adjustments not only in autonavi.com, but also Google maps (which I failed to update) and Bookmarks plugin. Instead of making a Monkey patch like this, it's better to set some variable to include the coordinate correction.

By the way, @magicbug2013 , I tried to update the coordinate fix for the new Google maps library, but the changes don't apply. https://github.com/IITC-CE/ingress-intel-total-conversion/tree/fix-google-map-offsets-in-china
Maybe you can fix it

johnd0e commented:
>  it's better to set some variable to include the coordinate correction.

...so if IITC knows that current map needs correction, than it begins correct coordinates for all plugins.
And it stops correction as soon as we select another map. Right?

modos189 commented:
Yes. The question is, should IITC have a list the maps that require adjustment, or should the plugins themselves call some function to enable

johnd0e commented:
Or may be it's possible to apply offset to map tiles only? In some way that coordinates stay intact.

modos189 commented:
Um. Wait, why change the coordinates of the Bookmarks? After all, correction is necessary only for map tiles...

magicbug2013 commented:
@modos189 
A few weeks ago, I tried to change the _update and _handleZoomAnim functions, as you did, but it didn't work, so now I'm trying to change the coordinates of protal, link and field. Now the problem is bookmarks. The best result is that don't change the coordinates of collection. just change the display position of the icon. I'll try it.

magicbug2013 committed:
fixed the coordinates and display of Bookmarks.

now Bookmarks begin to store the original coordinates and display them correctly.

modos189 commented:
Maybe we can make changes to the Leaflet to support correction?

johnd0e commented:
> Maybe we can make changes to the Leaflet to support correction?

In form of leaflet extension.
And at first we should check, may be such extension already exists, it's obvious that our project is not unique.

@magicbug2013, could you check if offset problem is solved here: https://github.com/htoooth/Leaflet.ChineseTmsProviders?

modos189 commented:
It can also be useful
https://github.com/hobo0cn/leaflet-gcj02

Perhaps this solution is for autonavi.com, but Google maps we connect in a completely different way

magicbug2013 commented:
> Maybe we can make changes to the Leaflet to support correction?

yes, we fixed the coordinates everywhere. In fact, we only need a function to move the map layer a little bit, but not any other layer.

magicbug2013 commented:
> It can also be useful
> https://github.com/hobo0cn/leaflet-gcj02
> 
> Perhaps this solution is for autonavi.com, but Google maps we connect in a completely different way



> > Maybe we can make changes to the Leaflet to support correction?
> 
> In form of leaflet extension.
> And at first we should check, may be such extension already exists, it's obvious that our project is not unique.
> 
> @magicbug2013, could you check if offset problem is solved here: https://github.com/htoooth/Leaflet.ChineseTmsProviders?

The fixed method can be referred to, but I don't know how to apply it to the project.

magicbug2013 commented:
> It can also be useful
> https://github.com/hobo0cn/leaflet-gcj02
> 
> Perhaps this solution is for autonavi.com, but Google maps we connect in a completely different way

He just added several maps of China and did not fix the coordinates.

modos189 commented:
> He just added several maps of China and did not fix the coordinates.

probably you are confused with
https://github.com/htoooth/Leaflet.ChineseTmsProviders

johnd0e commented:
Here is one more similar project: https://github.com/sureleo/leaflet-baidu

johnd0e commented:
@magicbug2013, also consider `Code Style` section of [HACKING.md](https://github.com/IITC-CE/ingress-intel-total-conversion/blob/master/HACKING.md#code-style).
> indent using two spaces

modos189 commented:
@johnd0e here we have to figure out how to move the map, not the portals. Compared to this, indents are a trifle)

johnd0e commented:
I propose to separate discussions.

> we have to figure out how to move the map, not the portals

As it is more general question, then let's create dedicated issue. I can do it, if you don't mind.

modos189 commented:
Ok

johnd0e commented on plugins/basemap-gaode-by-gmoogway.user.js:
```diff
@@ -0,0 +1,351 @@
+
+@@PLUGINSTART@@
+
+    // PLUGIN START ////////////////////////////////////////////////////////
```
Here should not be any indent, see other iitc plugins

magicbug2013 replied:
fixed

johnd0e commented on plugins/basemap-gaode-by-gmoogway.user.js:
```diff
@@ -0,0 +1,351 @@
+
+        var PluginBookmarksExist = false;
+        var PluginPlayerTrackerExist = false;
+        if (window.bootPlugins) {
```
This is wrong approach, because this code may be executed before that plugins loaded.

magicbug2013 replied:
fixed

magicbug2013 committed:
fixed the coordinates and display of draw-tools; fixed function injection

magicbug2013 commented:
@modos189 @johnd0e 
now the plugin works fine, maybe it can be merged?

modos189 commented:
@magicbug2013 Let's not override other plugins and change the coordinates of the portals.

[I was able to create](https://github.com/IITC-CE/ingress-intel-total-conversion/commit/64060c4b78564c5d0d7d40f0603b48560cbd01ac) a coordinate correction that only moves the map. So far it does not work perfectly, but it is much better than changing the coordinates of the portals.

I am sure we can bring this decision to perfection. Join the task:
https://github.com/IITC-CE/ingress-intel-total-conversion/issues/64#issuecomment-454941754

magicbug2013 commented:
> @magicbug2013 Let's not override other plugins and change the coordinates of the portals.
> 
> [I was able to create](https://github.com/IITC-CE/ingress-intel-total-conversion/commit/64060c4b78564c5d0d7d40f0603b48560cbd01ac) a coordinate correction that only moves the map. So far it does not work perfectly, but it is much better than changing the coordinates of the portals.
> 
> I am sure we can bring this decision to perfection. Join the task:
> [#64 (comment)](https://github.com/IITC-CE/ingress-intel-total-conversion/issues/64#issuecomment-454941754)

yes, that's better, It will be easier to add maps of China. I close this request, and re-fork the project.

johnd0e commented:
@magicbug2013 Wellcome to test #72 for Google offset.
And see there comment how to turn on offset correction for any map.

magicbug2013 commented:
@johnd0e 
great works.  there are no problems for maps using gcj02, such as Google map, Gaode map and so on, but there are still many problems for Baidu map (bd09), which is not only a problem of coordinate offset, but also the problem of coordinate system and Title map download path . I think it's possible to integrate gcj02 modifications into boot. js, leaving bd09 alone, and add an options attribute  fixChinaMapOffsetType = 'gcj02'.

johnd0e commented:
Have you set  `needFixChinaOffset: 'Baidu'` for Baidu?

> which is not only a problem of coordinate offset, but also the problem of coordinate system and Title map download path .

Do you have plugin for Baidu?

johnd0e commented:
BTW, do you know about these PRs:
https://github.com/iitc-project/ingress-intel-total-conversion/pull/1211
https://github.com/iitc-project/ingress-intel-total-conversion/pull/1262

~Are they usable?~
Have tried to fix them?

**Update**
Here is quick and dirty attempt to update the plugin: https://gist.github.com/johnd0e/4509cd02b71d60510be28ec7ffb8bd38
Please check if it works for you.
